### PR TITLE
No synced command

### PR DESCRIPTION
**Pull request by gfgtdf**

### CODE REVIEW

**Comment by gfgtdf** [comment]
This still needs some cleanups, in particular i don't know whre to put thos macros, maybe soemone elce could do that. also those custom commands (c2,c3) need better names

File: data/lua/custom_command.lua [hunk 1]
```diff
@@ -0,0 +1,16 @@
+local helper = wesnoth.require "helper"
+local T = helper.set_wml_tag_metatable {}
+
+local handlers = {}
+
+function wesnoth.game_events.on_sycned_command(cfg)
```
Comment (CelticMinstrel) [inline data/lua/custom_command.lua]: Misspelled, should be `on_synced_command`.

File: src/scripting/game_lua_kernel.cpp [hunk 2]
```diff
@@ -4395,6 +4395,17 @@ bool game_lua_kernel::run_event(const game_events::queued_event& ev)
 	return true;
 }
 
+void game_lua_kernel::custom_command(const config& cfg)
+{
+	lua_State *L = mState;
+
+	if (!luaW_getglobal(L, "wesnoth", "game_events", "on_sycned_command")) {
```
Comment (CelticMinstrel) [inline src/scripting/game_lua_kernel.cpp]: Misspelled, should be `on_synced_command`.

File: data/lua/custom_command.lua (hunk 1 above)
Comment (CelticMinstrel) [inline data/lua/custom_command.lua]: MIsspelled, should be `on_synced_command`.

File: src/scripting/game_lua_kernel.cpp (hunk 2 above)
Comment (CelticMinstrel) [inline src/scripting/game_lua_kernel.cpp]: MIsspelled, should be `on_synced_command`.

**Comment by gfgtdf** [comment]
>  if I understand correctly. It would (ideally) also allow wesnoth.synced_commands.name(cfg)


that would just call that function locally.

**Comment by CelticMinstrel** [comment]
I see your reasoning there, but it's definitely possible to do it so that that's not the case. Still, that part isn't as important; as long as registering the command is as simple as putting a function in a table, that would be acceptable to me.

You'd also eventually need to consider about what if they want to override a built-in synched command; but I think it'd be fine to not support that (treat keys that are the name of an existing command as read-only).

File: data/ai/utils/custom_command.cfg [hunk 3]
```diff
@@ -0,0 +1,18 @@
+
+#define ENABLE_AI_COMMAND_RABBIT_SPAWN
+[lua]
+code = <<
+
+wesnoth.register_synced_command("rabbit_despawn", function (cfg)
```
Comment (CelticMinstrel) [inline data/ai/utils/custom_command.cfg]: These should probably be registered when the MicroAI is enabled, which would mean somewhere in _data/ai/micro_ais/mai_defs/animals.lua_.

Comment (gfgtdf) [reply data/ai/utils/custom_command.cfg]: ye i know nothing about ai code, someone else can do that.

**Comment by CelticMinstrel** [comment]
I wonder what made you think it was acceptable to merge a PR that had a pending "changed requested" review.

**Comment by gfgtdf** [comment]
Well, i did implement your 'use a table instead of wesnoth.register_synced_command' suggestion.

**Comment by CelticMinstrel** [comment]
True, but you also broke the rabbits MicroAI and left in the old code in an #ifdef.

Admittedly I'd missed the new table when I looked before, so I'm not quite as annoyed about it anymore. The name is a bit long, though.

**Comment by gfgtdf** [comment]
> but you also broke the rabbits MicroAI 

didn't know how to test them, what exactly is borken?

> left in the old code in an #ifdef.

as i said in irc. today, we can probably remove more code than just the one i commented out, but since i don't know src/ai i left that part to othher people.

> The name is a bit long, though

hmm ye i know but i wanted it to make clear that it only contains the usder defined actions, not the normal ones like recruit/recall etc.

**Comment by CelticMinstrel** [comment]
As far as I can tell the rabbits macro is never used. Thus, broken.

About the name, I'd actually prefer if it _does_ contain built-in commands too, eventually.

**Comment by gfgtdf** [comment]
> As far as I can tell the rabbits macro is never used. Thus, broken.


well yes becasue no mainline scnario uses that ai afaik.

**Comment by CelticMinstrel** [comment]
Oh, and I understand you don't know about the AI and thus wanted to leave it to someone else, but in that case it would've been better to get someone else to push it to this branch _before_ merging it.

**Comment by mattsc** [comment]
All:  I just wanted to say sorry that I have been, and will continue to be for a while, entirely inactive on this.  I have a couple real-life issue to deal with that mean I neither have the time nor the mental calmness to look into this until early next week at the earliest.

**Pull request merged**